### PR TITLE
New `validate-blobs-sidecars` debug db subcommand

### PR DESCRIPTION
**Pull request by tbenr**
Adds consistency check command to validate block <-> blobs sidecars consistency in DB

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 1]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+    try (final Database database = createDatabase(beaconNodeDataOptions, eth2NetworkOptions)) {
+
+      // checking blobs for finalized blocks
+
+      Optional<SignedBeaconBlock> maybeFinalizedBlock = database.getLastAvailableFinalizedBlock();
+      maybeEarliestBlock = database.getEarliestAvailableBlock();
+      maybeEarliestBlobsSidecarSlot = database.getEarliestBlobsSidecarSlot();
+      maybeLatestBlobsSidecarSlot = maybeFinalizedBlock.map(SignedBeaconBlock::getSlot);
+      while (maybeFinalizedBlock.isPresent()
+          && maybeFinalizedBlock.get().getSlot().isGreaterThanOrEqualTo(UInt64.ZERO)) {
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: I don't understand when second condition could be not true. It's UInt

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 2]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+        SignedBeaconBlock currentBlock = maybeFinalizedBlock.get();
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: final

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 3]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+            // inside DA window but missing
+            missingFinalizedBlobsSidecars++;
+            System.err.printf(
+                "ERROR: Unable to locate finalized blobs sidecar %s%n",
+                slotAndBlockRoot.toLogString());
+          }
+        } else {
+          if (currentBlock.getSlot().isLessThan(lowerSlotBoundDataAvailability)) {
+            // outside DA window but exists
+            blobsSidecarsWithoutBlocks++;
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: Are we increasing the correct counter? It doesn't match the message printed

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 4]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+    Optional<UInt64> maybeLatestBlobsSidecarSlot = Optional.empty();
+    Optional<SignedBeaconBlock> maybeEarliestBlock = Optional.empty();
+    try (final Database database = createDatabase(beaconNodeDataOptions, eth2NetworkOptions)) {
+
+      // checking blobs for finalized blocks
+
+      Optional<SignedBeaconBlock> maybeFinalizedBlock = database.getLastAvailableFinalizedBlock();
+      maybeEarliestBlock = database.getEarliestAvailableBlock();
+      maybeEarliestBlobsSidecarSlot = database.getEarliestBlobsSidecarSlot();
+      maybeLatestBlobsSidecarSlot = maybeFinalizedBlock.map(SignedBeaconBlock::getSlot);
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: Could `maybeLatestFinalizedBlobsSidecarSlot` name better indicate its purpose?

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 5]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+            } else {
+              // finalized
+              final Optional<SignedBeaconBlock> maybeFinalizedBlockFound =
+                  database.getFinalizedBlockAtSlot(blobsSidecarsKey.getSlot());
+              if (maybeFinalizedBlockFound.isPresent()) {
+                if (maybeFinalizedBlockFound
+                    .get()
+                    .getRoot()
+                    .equals(blobsSidecarsKey.getBlockRoot())) {
+                  blobsSidecarsWithoutBlocks++;
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: Same here. It looks like unpruned counter, maybe just not talking name?

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 6]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+          final boolean isUnconfirmed = unconfirmedBlobsSidecars.contains(blobsSidecarsKey);
+
+          if (isUnconfirmed) {
+            // unconfirmed, check a corresponding hot or finalized block doesn't exist
+            if (blobsSidecarsKey
+                .getSlot()
+                .isGreaterThan(maybeLatestBlobsSidecarSlot.orElse(UInt64.ZERO))) {
+              // hot
+              if (database.getHotBlock(blobsSidecarsKey.getBlockRoot()).isPresent()) {
+                blobsSidecarsWithoutBlocks++;
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: Same here. It looks like unpruned counter, maybe just not self-speaking name?

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 7]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+                System.err.printf(
+                    "ERROR: unconfirmed blobs sidecar has a corresponding hot block %s%n",
+                    blobsSidecarsKey.toLogString());
+              }
+            } else {
+              // finalized
+              final Optional<SignedBeaconBlock> maybeFinalizedBlockFound =
+                  database.getFinalizedBlockAtSlot(blobsSidecarsKey.getSlot());
+              if (maybeFinalizedBlockFound.isPresent()) {
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: what if not? do we always expect to find it from other side iteration (by block)?

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 8]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+          if (isUnconfirmed) {
+            // unconfirmed, check a corresponding hot or finalized block doesn't exist
+            if (blobsSidecarsKey
+                .getSlot()
+                .isGreaterThan(maybeLatestBlobsSidecarSlot.orElse(UInt64.ZERO))) {
+              // hot
+              if (database.getHotBlock(blobsSidecarsKey.getBlockRoot()).isPresent()) {
+                blobsSidecarsWithoutBlocks++;
+                System.err.printf(
+                    "ERROR: unconfirmed blobs sidecar has a corresponding hot block %s%n",
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: if it has even counter for it, maybe it's WARNING?

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 9]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+          System.err.printf(
+              "ERROR: Unable to locate unconfirmed blobs sidecar in main blobs sidecar table %s%n",
+              slotAndBlockRoot.toLogString());
+        }
+      }
+    } catch (DatabaseStorageException ex) {
+      System.out.println("Failed to open database");
+    }
+
+    final long endTime = System.currentTimeMillis();
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: endTimeMillis for consistency?

File: teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java [hunk 10]
```diff
@@ -439,6 +447,290 @@ public int dumpBlobsSidecar(
...
+        "\tCalculated Data Availability window (slots): %s - %s%n",
+        lowerSlotBoundDataAvailability, currentSlot);
+    System.out.printf("\tEarliest BlobsSidecar slot: %s%n", maybeEarliestBlobsSidecarSlot);
+    System.out.printf(
+        "\tEarliest finalized block slot: %s%n",
+        maybeEarliestBlock.map(SignedBeaconBlock::getSlot));
+    System.out.printf("\tLatest finalized block slot: %s%n", maybeLatestBlobsSidecarSlot);
+    System.out.printf("\tMissing finalized BlobsSidecars: %s%n", missingFinalizedBlobsSidecars);
+    System.out.printf("\tMissing hot BlobsSidecars: %s%n", missingHotBlobsSidecars);
+    System.out.printf("\tExcess BlobsSidecars: %s%n", blobsSidecarsWithoutBlocks);
```
Comment (zilm13) [inline teku/src/main/java/tech/pegasys/teku/cli/subcommand/debug/DebugDbCommand.java]: Ok, I see, it's some total of unpruned from other side. I think it's just needs better name

**Comment by tbenr** [comment]
yeah let me revisit them, I implemented the command quite quickly and starting by simply copying the existing validate command.

**Review by zilm13 — APPROVED** [review]
LGTM

**Pull request merged**